File: solutions/backtracking/alternatives/047-permutations-ii.js
/**
 * 47. Permutations Ii
 * Medium
 *
 * Permutations II - Find all unique permutations in an array that may contain duplicates @param {number[]} nums - Input array of numbers (may contain duplicates) @return {number[][]} - Array of all unique permutations
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Permutations Ii is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Permutations II - Find all unique permutations in an array that may contain duplicates
 * 
 * @param {number[]} nums - Input array of numbers (may contain duplicates)
 * @return {number[][]} - Array of all unique permutations
 */
function permuteUnique(nums) {
    // Edge case: empty array
    if (!nums || nums.length === 0) {
        return [];
    }

    // Sort the array first to handle duplicates efficiently
    nums.sort((a, b) => a - b);
    const result = [];
    const used = new Array(nums.length).fill(false);
    
    /**
     * Backtracking helper function to generate permutations
     * @param {number[]} current - Current permutation being built
     */
    function backtrack(current) {
        // Base case: if current permutation length equals input array length
        if (current.length === nums.length) {
            result.push([...current]);
            return;
        }

        for (let i = 0; i < nums.length; i++) {
            // Skip if number is already used in current permutation
            if (used[i]) continue;

            // Skip duplicates to avoid duplicate permutations
            if (i > 0 && nums[i] === nums[i - 1] && !used[i - 1]) continue;

            // Include current number in permutation
            used[i] = true;
            current.push(nums[i]);

            // Recursive call to generate next number in permutation
            backtrack(current);

            // Backtrack: remove current number and mark as unused
            current.pop();
            used[i] = false;
        }
    }

    // Start backtracking with empty permutation
    backtrack([]);
    return result;
}

// Example usage and test cases
function runTests() {
    const testCases = [
        [1, 1, 2],
        [1, 2, 3],
        [1],
        [1, 1, 1],
        []
    ];

    console.log("Running test cases for permuteUnique:");
    testCases.forEach((test, index) => {
        console.log(`\nTest case ${index + 1}:`);
        console.log("Input:", test);
        console.log("Output:", permuteUnique(test));
    });
}

// Export the function for potential module usage
module.exports = {
    permuteUnique
};

// Run tests if not being imported as a module
if (require.main === module) {
    runTests();
}

File: docs/solutions/backtracking/047-permutations-ii.py
"""
# Difficulty: Medium

# 047. Permutations Ii

Given a problem that demonstrates key concepts in Backtracking.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[input description]</dd>
<dt>Output:</dt>
<dd>[output description]</dd>
<dt>Explanation:</dt>
<dd>[explanation]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of backtracking concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply backtracking methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages backtracking principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        """
        Generate all possible unique permutations of the input array.
        
        Args:
            nums: List of integers (may contain duplicates)
            
        Returns:
            List of lists containing all unique permutations
        
        Time Complexity: O(n!)
        Space Complexity: O(n)
        """
        # Handle empty input
        if not nums:
            return []
        
        def backtrack(counter: Counter, temp_perm: List[int], n: int) -> None:
            """
            Helper function for backtracking to generate permutations.
            
            Args:
                counter: Counter object containing remaining numbers
                temp_perm: Current permutation being built
                n: Target length of permutation
            """
            # Base case: if current permutation is complete
            if len(temp_perm) == n:
                result.append(temp_perm[:])
                return
            
            # Try each unique number from counter
            for num in counter:
                if counter[num] > 0:
                    # Add current number to permutation
                    temp_perm.append(num)
                    counter[num] -= 1
                    
                    # Recursive call
                    backtrack(counter, temp_perm, n)
                    
                    # Backtrack
                    temp_perm.pop()
                    counter[num] += 1
        
        result = []
        # Create counter for frequency of each number
        counter = Counter(nums)
        backtrack(counter, [], len(nums))
        return result

def test_solution():
    """
    Test cases for 047. Permutations Ii.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 047. Permutations Ii")
